### ISSUE TITLE
Support StarkNet state `copy`?

### ISSUE THREAD

**Issue by RoboTeddy** [Shard-Labs/starknet-devnet#9]
`StarkNet` (from `starkware.starknet.testing.starknet`) has a `copy` method: https://github.com/starkware-libs/cairo-lang/blob/577d7a5ca085fc76fb1cb18e04d94991127e3724/src/starkware/starknet/testing/state.py#L51

The copy method is super useful for writing efficient tests, because you can deploy/initialize/etc a bunch of inter-related contracts, and then make multiple copies of that StarkNet state in order to run various tests on them in parallel.

I love that `starknet-hardhat-plugin` makes it easy to write tests against a local `starknet-devnet` instance — however I find myself missing the efficiency of the `copy` method.

Would you be open to supporting this functionality inside `starknet-devnet`? We could add a `/gateway/copy` endpoint which returns a new identifier which later could be optionally passed as a header or request arg to various gateway endpoints such as `/gateway/add_transaction`. (Of course, if this header/arg were omitted, the request would apply to the root starknet instance.)

**Comment by RoboTeddy** [comment]
(P.S. If you're open to this feature, I may end up finding time to contribute it!)

**Comment by FabijanC** [comment]
Thank you for sharing the idea and starting this discussion @RoboTeddy!

I'm aware of the `copy` method. In fact, it's used inside Starknet's `call`, allowing it to first copy the state and then works on the copy, leaving the original intact. Also I'm not sure how much this is related to your use-case, but this text from [here](https://www.cairo-lang.org/docs/hello_starknet/intro.html#query-the-balance) might be related:
> using call instead of invoke on a function that may change the state, such as increase_balance, will return the result of the function without actually applying it to the current state, allowing the user to dry-run before committing to a state update.

Are you talking about using the devnet from `starknet-hardhat-plugin` or by using Starknet CLI directly? Or are you usage-agnostic, just discussing general functionalities of the devnet? Because I'm not sure how the state instance identifier would be provided?

Also to be clear, the API provided by the devnet is basically a copy of the one provided by Alpha testnet. Although new endpoints could be added, sure, but who would be able to use them?

**Comment by RoboTeddy** [comment]
Thanks for being so responsive @FabijanC !

> Are you talking about using the devnet from starknet-hardhat-plugin
> Although new endpoints could be added, sure, but who would be able to use them?

The case I had in mind was for writing tests using `starknet-hardhat` as in [these examples](https://github.com/Shard-Labs/starknet-hardhat-example/blob/master/test/sample-test.ts).

Perhaps after tests that do the initial deployments of various contracts, subsequent tests could each opt to copy that initial deployment and test against the copy. In total it'd require:
- modifying both `starknet-devnet` to add support for copies
- modifying `starknet-hardhat-plugin` to add support for using copies

**Comment by FabijanC** [comment]
This might require more changes (maybe a whole paradigm change) to the existing plugin code, because currently there is no direct interaction with the testnet you're interacting with. And there is also no testnet abstraction (like a class representing the testnet/devnet) which would enable you to perform the copying.

**Comment by RoboTeddy** [comment]
OK! Feel free to close this issue if it seems out of scope for the project — I definitely do appreciate projects that keep things simple and as a result are more stable :)
